src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -51,19 +64,19 @@ internal class CodeRefactoringService : ICodeRefactoringService
             }
         }
 
-        private ImmutableDictionary<string, Lazy<IEnumerable<CodeRefactoringProvider>>> LanguageToProvidersMap
+        private ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>> LanguageToProvidersMap
             => _lazyLanguageToProvidersMap.Value;
 
-        private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
+        private ImmutableArray<CodeRefactoringProvider> GetProviders(Document document)
         {
+            var allRefactorings = ImmutableArray<CodeRefactoringProvider>.Empty;
             if (this.LanguageToProvidersMap.TryGetValue(document.Project.Language, out var lazyProviders))
             {
-                return lazyProviders.Value;
-            }
-            else
-            {
-                return SpecializedCollections.EmptyEnumerable<CodeRefactoringProvider>();
+                allRefactorings = lazyProviders.Value;
             }
+
+            allRefactorings = allRefactorings.AddRange(GetProjectRefactorings(document.Project));
+            return allRefactorings;
```

Comment:
any concern about this being called a lot and causing lots of allocations?

Reply:
should we invert ths.  specifically since we need to check at the project jeve, we just cache the result at the project level *including* the global set of refactorings.

Reply:
I agree with @CyrusNajmabadi - I recall we hit the exact perf issue that Cyrus mentions here in CodeFixService and added two level cache [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/CodeFixes/CodeFixService.cs,697) and [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/CodeFixes/CodeFixService.cs,706)

Reply:
Fast-path allocations should now be eliminated, and slow-path allocations substantially reduced.